pkg/upgrade/machine_deployment.go

```diff
@@ -22,15 +26,55 @@ func NewMachineDeploymentUpgrader(log logr.Logger, config Config) (*MachineDeplo
 		return nil, errors.Wrap(err, "error initializing upgrader")
 	}
 
+	var selector labels.Selector
+	if config.MachineDeployment.LabelSelector != "" {
+		// Step 1: parse the user-specified label selector
+		selector, err = labels.Parse(config.MachineDeployment.LabelSelector)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error parsing machine deployment label selector %q", config.MachineDeployment.LabelSelector)
+		}
+
+		// Step 2: add the cluster name to the label selector to avoid the possibility of selecting machine deployments
+		// that belong to clusters other than the one the user specified.
+		r, err := labels.NewRequirement(clusterv1.MachineClusterLabelName, selection.Equals, []string{b.clusterName})
+		if err != nil {
+			return nil, errors.Wrap(err, "error adding cluster name to label selector")
+		}
+
+		selector = selector.Add(*r)
+	}
+
 	return &MachineDeploymentUpgrader{
-		base: b,
+		base:     b,
+		name:     config.MachineDeployment.Name,
+		selector: selector,
 	}, nil
 }
 
 func (u *MachineDeploymentUpgrader) Upgrade() error {
-	machineDeployments, err := u.listMachineDeployments()
-	if err != nil {
-		return err
+	var (
+		machineDeployments *clusterv1.MachineDeploymentList
+		err                error
+	)
+
+	if u.name != "" {
+		key := ctrlclient.ObjectKey{
+			Namespace: u.clusterNamespace,
+			Name:      u.name,
+		}
+
+		var machineDeployment clusterv1.MachineDeployment
+
+		if err := u.managerClusterClient.Get(context.TODO(), key, &machineDeployment); err != nil {
+			return errors.Wrapf(err, "error getting machine deployment %q", u.name)
+		}
+
+		machineDeployments = &clusterv1.MachineDeploymentList{Items: []clusterv1.MachineDeployment{machineDeployment}}
```

Comment:
Should this validate that the returned machineDeployment belongs to u.ClusterName?

Reply:
Will do